### PR TITLE
[SPARK-24193] create TakeOrderedAndProjectExec only when the limit number is below spark.sql.execution.topKSortFallbackThreshold.

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

Physical plan of `select colA from t order by colB limit M` is `TakeOrderedAndProject`;
Currently `TakeOrderedAndProject` sorts data in memory, see https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala#L158 
We can add a config – if the number of limit (M) is too big, we can sort by disk. Thus memory issue can be resolved.

## How was this patch tested?

Test added

### CODE REVIEW

**Comment by jinxing64** [comment]
cc @cloud-fan @viirya 
I'm not sure if you are interested in this config.
Could you please give some advice? 
Thanks a lot !

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 1]
```diff
@@ -127,13 +127,36 @@ case class TakeOrderedAndProjectExec(
...
+    }
+  }
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = {
+    if (limit < sortInMemThreshold) {
+      super.requiredChildOrdering
+    } else {
+      Seq(sortOrder)
+    }
+  }
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: It was only shuffling limit-ed rows from each partitions. The final sort was also only applied on limited-data form all partitions.

Now it requires sorting and shuffling on whole data. I suspect that it'd degrade performance in the end.

Comment (jinxing64) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: Thanks a lot for comments :)
Our users have queries with big number of limit, which suffers OOM with current code. So I added a config, below which we go by current code, if not we do global sort.
Yes there will be performance degradation. But it's better than OOM;
I think a limit-ed heap sort by disk will be a better idea. But I'm didn't find an existing implementation and don't want to make it too complicated here.
Thanks again for review this @viirya

**Comment by viirya** [comment]
Instead of touching inside of `TakeOrderedAndProjectExec`, how about we don't replace `Sort` + `Limit` with `TakeOrderedAndProjectExec` when reaching the threshold? A.k.a:

```scala
  object SpecialLimits extends Strategy {
    override def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
      case ReturnAnswer(rootPlan) => rootPlan match {
        case Limit(IntegerLiteral(limit), Sort(order, true, child)) if limit < conf.sortInMemForLimitThreshold =>
          TakeOrderedAndProjectExec(limit, order, child.output, planLater(child)) :: Nil
         ...
  }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1238,6 +1238,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val SORT_IN_MEM_FOR_LIMIT_THRESHOLD =
+    buildConf("spark.sql.limit.sortInMemThreshold")
+      .internal()
+      .doc("In sql like 'select x from t order by y limit m', if m is under this threshold, " +
+          "sort in memory, otherwise do a global sort with disk.")
+      .intConf
+      .createWithDefault(2000)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: A numeric limit threshold seems to hard to determine, how about a boolean config (default true) we can enable `TakeOrderedAndProjectExec` or not? cc @cloud-fan

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: what if users only have a few queries which have large limit and they want to disable the top n sort? I feel this config is more flexible than a boolean flag.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -1238,6 +1238,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val SORT_IN_MEM_FOR_LIMIT_THRESHOLD =
+    buildConf("spark.sql.limit.sortInMemThreshold")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `spark.sql.execution.combineLimitAfterSortTreshold`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -1238,6 +1238,14 @@ object SQLConf {
...
+      .internal()
+      .doc("In sql like 'select x from t order by y limit m', if m is under this threshold, " +
+          "sort in memory, otherwise do a global sort with disk.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `with disk` -> `which spills to disk if necessary`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Oh, yeah, reasonable.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Isn't 2000 too small for this?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I would suggest `Int.Max` as the default value, which preserves the previous behavior. Users can tune it w.r.t. their workload.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Yeah, I agree.

**Comment by jinxing64** [comment]
@cloud-fan @viirya 
Thanks for comments. I refined accordingly. Please check~

**Comment by jiangxb1987** [comment]
The changes looks good to me, but it should also be great to have a test suite to cover this change. Seems we don't have a test suite for the rule `SpecialLimits` ?

**Comment by viirya** [comment]
We can have a test to check the physical plan under different threshold.

**Comment by jinxing64** [comment]
I will add a suite tomorrow.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Review by viirya — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -1238,6 +1238,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val COMBINE_LIMIT_AFTER_SORT_THRESHOLD =
+    buildConf("spark.sql.execution.combineLimitAfterSortThreshold")
+      .internal()
+      .doc("In sql like 'select x from t order by y limit m', if m is under this threshold, " +
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `In sql queries with a sort followed by a limit like 'select x from t order by y limit m', if m ...`

**Comment by jiangxb1987** [comment]
retest this please

**Comment by viirya** [comment]
retest this please.

**Comment by jinxing64** [comment]
PR description updated.

**Comment by viirya** [comment]
Thanks! LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -1238,6 +1238,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val COMBINE_LIMIT_AFTER_SORT_THRESHOLD =
+    buildConf("spark.sql.execution.combineLimitAfterSortThreshold")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: after some more thoughts, this config name is not very obvious. How about `topNSortFallbackThreshold`? any better ideas?

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: topK

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: or even simpler, `topKThreshold`

Comment (kiszk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `topK` seems to be selected from many samples. 
How about `limitKSortFallbackThreshold` or `limitKThreshold`?

Comment (jinxing64) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I go by `topKSortFallbackThreshold` and updated the change.

**Comment by kiszk** [comment]
retest this please

**Comment by cloud-fan** [comment]
can you change the PR title? otherwise LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**